object/policy_mapper.go

```diff
@@ -0,0 +1,400 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package object
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"runtime/debug"
+	"strings"
+
+	"github.com/casdoor/casdoor/util"
+	"golang.org/x/sync/errgroup"
+)
+
+type policyDomain struct {
+	id        string
+	name      string
+	subDomain string
+}
+
+type policyGroup struct {
+	id          string
+	name        string
+	parentGroup string
+	user        string
+}
+
+type policyRole struct {
+	id      string
+	name    string
+	domain  policyDomain
+	user    string
+	group   policyGroup
+	subRole string
+
+	empty bool
+}
+
+type policyPermission struct {
+	id       string
+	domain   policyDomain
+	resource string
+	action   string
+	role     policyRole
+	user     string
+	group    policyGroup
+	effect   string
+
+	empty bool
+}
+
+func getValueByItem(policyPermissionItem policyPermission, policyItem string) (string, bool) {
+	if policyItem == "" {
+		return policyItem, true
+	}
+
+	if policyItem[:5] == "const" {
+		return policyItem[6:], true
+	}
+
+	var value string
+
+	switch policyItem {
+	case "role.name":
+		value = policyPermissionItem.role.name
+	case "role.subrole":
+		value = policyPermissionItem.role.subRole
+	case "role.domain.name":
+		value = policyPermissionItem.role.domain.name
+	case "role.domain.subdomain":
+		value = policyPermissionItem.role.domain.subDomain
+	case "role.user":
+		value = policyPermissionItem.role.user
+	case "role.group.name":
+		value = policyPermissionItem.role.group.name
+	case "role.group.parentgroup":
+		value = policyPermissionItem.role.group.parentGroup
+	case "role.group.user":
+		value = policyPermissionItem.role.group.user
+	case "permission.action":
+		value = policyPermissionItem.action
+	case "permission.resource":
+		value = policyPermissionItem.resource
+	case "permission.user":
+		value = policyPermissionItem.user
+	case "permission.effect":
+		value = policyPermissionItem.effect
+	case "permission.domain.name":
+		value = policyPermissionItem.domain.name
+	case "permission.domain.subdomain":
+		value = policyPermissionItem.domain.subDomain
+	case "permission.group.name":
+		value = policyPermissionItem.group.name
+	case "permission.group.parentgroup":
+		value = policyPermissionItem.group.parentGroup
+	case "permission.group.user":
+		value = policyPermissionItem.group.user
+	}
+
+	if value == "" {
+		return value, false
+	}
+
+	return value, true
+}
+
+func refillWithEmptyStrings(row []string) []string {
+	for len(row) < 7 {
```

Comment:
move to dedicated helper / string utils

Reply:
[]string replaced by [7]string, so function is removed now.